README.md
# iconfont-builder

[中文版](https://github.com/q-iconfont/iconfont-builder/blob/master/README_CN.md)

## Introduction

Iconfont-builder is a node.js package for providing a middleware that create some font files.

## Installation (via [npm](https://npmjs.org/package/iconfont-builder))

```bash
$ npm i --save iconfont-builder
```

## Usage

### Simple Usage

```js
var builder = require('iconfont-builder');
var path = require('path');

var options = {
    icons: [
        {
            name: 'www-font-o',
            file: 'abc.svg',
            codepoint: 61441
        }
    ],
    src: path.join(__dirname, 'src'),
    fontName: 'iconfont',
    descent: 0,
    dest: path.join(__dirname, 'dest')
};

builder(options)
    .then().catch();
```

### List of options

#### icons

Type: `Array<Object>`

Example:

```js
{
    name: 'www-font-o', // className of icon
    file: 'abc.svg',    // fileName of icon
    codepoint: 61441    // unicode of icon
}
```

#### writeFiles

Type: `Boolean`

Default: `true`

It is possible to not create font files but get the attribute **d** of each icon svg. The attribute d contains all paths' information of an icon, which can be use to draw a svg icon.

#### readFiles

Type: `Boolean`

Default: `true`

You can only use attribute **d** to create font files! If this param is `false`, the Object in param `icons` should have attribute `d`.

#### fontName

Type: `String`

Default: `'iconfont'`

Name of font and font files.

#### startCodePoint

Type: `Number`

Default: `0xF000`

Start of font's unicode in DEC(e.g. `61441`) or HEX(e.g. `0xF001`). When passing `options` without `icons`, builder will use `startCodePoint` as the first unicode of font icon, and the unicode of each remaining icons will increased by one in order.

#### src

Type: `String`

Default: `'.'`

Directory of source svg font files.

#### dest

Type: `String`

Directory for generated font files.

#### descent

Type: `Number`

Default: `0`

The font descent. It's useful to fix the font baseline yourself.

Warning: The descent is a positive value!

## Author

[missmiss](http://www.weibo.com/ssherrylliu)

**[Follow me (@missmiss) on Github!](http://missmiss.github.io/)**


HISTORY.md
1.0.1 / 2015-07-01
------------------

- Added examples
- Multiple fixes


1.0.0 / 2015-06-26
------------------

- First release


test/parser/_index.js
var _ = require('underscore');

var builder = require('../../lib');

var basePath = 'test/parser';

function getOptions() {
  return {
    icons: [
      {
        name: 'www-font-o',
        codepoint: 0xF001,
        file: 'test.svg'
      }
    ],
    src: basePath,
    fontName: 'iconfont'
  };
}

describe('能正确解析字体文件', function() {
  it('不生成字体，获取path信息', function(done) {
    var options = getOptions();
    options.writeFiles = false;

    builder(options)
      .then(function(icons) {
        if (!icons) done(new Error('no icons'));
        _.each(icons, function(icon) {
          if (!icon.d) {
            done(new Error('icon ' + icon.name + 'has no path'));
          }
          console.log(icon.d);
        });
        done();
      })
  });
});

test/builder/_index.js
var assert = require('assert');
var fs = require('fs');
var path = require('path');
var Q = require('q');

var _ = require('underscore');
var builder = require('../../lib');

var basePath = 'test/builder';
var dest = path.join(basePath, 'dest');

function getOptions() {
  return {
    icons: [
      {
        name: 'www-font-o',
        codepoint: 0xF001,
        file: 'test.svg'
      }
    ],
    src: basePath,
    dest: dest,
    fontName: 'iconfont'
  };
}

function generateFonts(options) {
  var hasDirQ = Q.nfcall(fs.access, dest);
  var rmDirQ = Q.nfcall(fs.rmdir, dest);

  return hasDirQ
    .then(rmDirQ)
    .finally(function() {
      return builder(options);
    })
    .then(function() {
      var fonts = ['svg', 'ttf', 'eot', 'woff'];
      var fontsQ = _.map(fonts, function(type) {
        var file = path.join(dest, 'iconfont.' + type);
        return Q.nfcall(fs.access, file);
      });

      return Promise.all(fontsQ)
    });
}

describe('生成正确的字体文件', function() {
  it('传递完整的icon信息', function(done) {
    var options = getOptions();

    generateFonts(options)
      .then(function() {
        done();
      })
      .catch(function(err) {
        done(err);
      });
  });

  it('缺失部分icon信息', function(done) {
    var options = getOptions();
    delete options.icons[0].name;

    generateFonts(options)
      .then(function() {
        done(new Error('It cannot be resolved'))
      })
      .catch(function() {
        done();
      });
  });

  it('不传递icon信息', function(done) {
    var options = getOptions();
    delete options.icons;

    generateFonts(options)
      .then(function() {
        done();
      })
      .catch(function(err) {
        done(err);
      });
  });

  it('使用d数据来生成字体', function(done) {
    var options = getOptions();
    delete options.icons[0].file;
    options.icons[0].d = 'M828.8 387.2C744.2 302.6 631.6 256 512 256C392.4 256 279.8 302.6 195.2 387.2S64 584.4 64 704V768C64 785.6 78.4 800 96 800H928C945.6 800 960 785.6 960 768V704C960 584.4 913.4 471.8 828.8 387.2zM896 736H128V704C128 492.2 300.2 320 512 320S896 492.2 896 704V736zM928 832H96C78.4 832 64 846.4 64 864S78.4 896 96 896H928C945.6 896 960 881.6 960 864S945.6 832 928 832zM512 224C538.4 224 560 202.4 560 176S538.4 128 512 128C485.6 128 464 149.6 464 176S485.6 224 512 224zM512 160C520.8 160 528 167.2 528 176S520.8 192 512 192S496 184.8 496 176S503.2 160 512 160zM352 400C352 408.837 359.163 416 368 416C376.837 416 384 408.837 384 400C384 391.163 376.837 384 368 384C359.163 384 352 391.163 352 400zM302.4 421.2C217 484.6 165.4 581.4 160.4 687.2C160 696 166.8 703.6 175.6 704C175.8 704 176.2 704 176.4 704C184.8 704 192 697.4 192.4 688.8C197 592.6 244 504.4 321.6 447C328.6 441.8 330.2 431.8 325 424.6C319.6 417.4 309.6 415.8 302.4 421.2z';
    options.readFiles = false;

    generateFonts(options)
      .then(function() {
        done();
      })
      .catch(function(err) {
        done(err);
      });
  });
});


lib/fontGenerator.js
/**
 * 生成eot, svg, ttf, woff 四种格式的文件
 * 支持只返回各个字体的 path-d 而不生成字体文件
 * 支持通过仅传入 d 来生成字体
 */

var fs = require('fs');
var path = require('path');

var _ = require('underscore');
var s2s = require('svgicons2svgfont');
var svg2ttf = require('svg2ttf');
var ttf2eot = require('ttf2eot');
var ttf2woff = require('ttf2woff');

var parser = require('./svgFontParser');

/**
 * 生成 svg 字体
 *
 * @param {Array<Object>} icons 图标数组，包含图标类名与 code
 * @param {Object} svgOpts s2s 插件的参数对象
 * @param {Object} options iconfont-builder 的参数对象
 * @returns {Promise}
 */
function generateSvg(icons, svgOpts, options) {
  var stream;
  var font = new Buffer(0);
  var def = Promise.defer();

  // 进行内部自定义设定
  svgOpts.normalize = true;   // 大小统一
  svgOpts.fontHeight = 1024;  // 高度统一为1024
  svgOpts.round = 1000;       // path值保留三位小数
  svgOpts.log = function(){}; // 沉默控制台输出

  stream = s2s(svgOpts)
    .on('data', function(data) {
      font = Buffer.concat([font, data]);
    })
    .on('error', function(err) {
      def.reject(err);
    })
    .on('finish', function() {
      def.resolve(font.toString());
    });

  _.each(icons, function(icon) {
    try {
      var iconFile = path.join(options.src, icon.file);
      var glyph = fs.createReadStream(iconFile);
      glyph.metadata = {
        name: icon.name,
        unicode: [String.fromCharCode(icon.codepoint)]
      };
      stream.write(glyph);
    } catch(e) {
      def.reject(e);
      return false;
    }
  });

  stream.end();
  return def.promise;
}

/**
 * 生成 ttf 字体，依赖于 svg 字体的生成
 *
 * @param {String} svgFont
 */
function generateTtf(svgFont) {
  var font = svg2ttf(svgFont);
  return new Buffer(font.buffer);
}

/**
 * 生成 woff 字体，依赖于 ttf 字体的生成
 *
 * @param {String} ttfFont
 */
function generateWoff(ttfFont) {
  var font = ttf2woff(new Uint8Array(ttfFont));
  return new Buffer(font.buffer);
}

/**
 * 生成 eot 字体，依赖于 ttf 字体的生成
 *
 * @param {String} ttfFont
 */
function generateEot(ttfFont) {
  var font = ttf2eot(new Uint8Array(ttfFont));
  return new Buffer(font.buffer);
}

/**
 * 按照依赖关系生成字体
 *
 * @param {Object} options 生成字体参数对象
 * @returns {Promise}
 */
function generateFonts(options) {
  // 使用 ascent 和 descent 进行字体的基线调整
  var svgOpts = _.pick(options,
    'fontName', 'ascent', 'descent'
  );
  var svg;

  // 首先进行 svg 的生成
  // 当 readFiles 为 false 时，使用 d 生成
  // 否则使用文件生成
  if (!options.readFiles) {
    svg = parser.getSvgIcon(options);
  } else {
    svg = generateSvg(options.icons, svgOpts, options);
  }

  // ttf 依赖 svg 的生成
  var ttf = svg.then(generateTtf);
  // eot 和 woff 依赖 tff 的生成
  var eot = ttf.then(generateEot);
  var woff = ttf.then(generateWoff);

  return Promise.all([svg, ttf, eot, woff]);
}

module.exports = generateFonts;


lib/svgFontParser.js
var _ = require('underscore');
var path = require('path');
var fs = require('fs');
var Q = require('q');
var handlebars = require('handlebars');

var xml = require('xmldom');
var Parser = xml.DOMParser;

// 如果有可能，还是使用 sax，不吃内存
function getPathData(svgContent, options) {
  var doc = new Parser().parseFromString(svgContent, 'text/xml');
  var icons = options.icons;
  var glyphs = doc.getElementsByTagName('glyph');
  var i = 0, len = glyphs.length;

  for (; i < len; i++) {
    var glyph = glyphs[i];
    var name = glyph.getAttribute('glyph-name');
    var d = glyph.getAttribute('d');

    _.each(icons, function(icon) {
      if (icon.name === name) {
        icon.d = d;
        return false;
      }
    });
  }

  return icons;
}

function getSvgIcon(options) {
  var tmpPath = path.join(__dirname, 'svg.handlebars');
  return Q.nfcall(fs.readFile, tmpPath, 'utf-8')
    .then(function(source) {
      var template = handlebars.compile(source);
      return template(options);
    });
}

exports.getPathData = getPathData;
exports.getSvgIcon = getSvgIcon;

lib/index.js
var fs = require('fs');
var path = require('path');
var Q = require('q');
var _ = require('underscore');
var mkdirp = require('mkdirp');

var generator = require('./fontGenerator');
var parser = require('./svgFontParser');

var DEFAULT_OPTIONS = {
  readFiles: true,
  writeFiles: true,
  fontName: 'iconfont',
  startCodePoint: 0xF000,
  src: '.',
  dest: '.',
  descent: 0
};

/**
 * 入口函数，根据参数生成字体或 icon 对象
 *
 * @param {Object} options 传递参数，详见 readme
 * @returns {Promise}
 */
function builder(options) {
  options = _.extend({}, DEFAULT_OPTIONS, options);
  options.ascent = 1024 - options.descent;

  // 填充 icons 数据
  return fillIcons(options)
    .then(function(icons) {
      options.icons = icons;
      return generator(options);
    })
    .then(function(fonts) {
      if (options.writeFiles) {
        return writeFonts(fonts, options);
      } else {
        // 直接返回包含 d 的 icon 数据
        return parser.getPathData(fonts[0], options);
      }
    });
}

/**
 * 字体写入方法，生成四种字体
 *
 * @param {Array<String>} fonts 字体内容数组
 * @param {Object} options 参数对象
 * @returns {Promise}
 */
function writeFonts(fonts, options) {
  var type = ['svg', 'ttf', 'eot', 'woff'];

  var fontsQ = _.map(fonts, function(font, i) {
    var filePath = path.join(options.dest, options.fontName + '.' + type[i]);

    var mkdirQ = Q.nfcall(mkdirp, path.dirname(filePath));
    var writeFileQ = Q.nfcall(fs.writeFile, filePath, font);

    return mkdirQ.then(writeFileQ);
  });

  return Promise.all(fontsQ);
}

/**
 * 判断是否传入 icons 对象，选择排查或补充
 *
 * @param {Object} options 参数对象
 * @returns {Promise}
 */
function fillIcons(options) {
  // 如果有 icons 数据，确保数据不为空
  if (options.icons) {
    var def = Promise.defer();

    _.each(options.icons, function(icon) {
      // name 和 codepoint 是必备的
      if (!icon.codepoint || !icon.name) {
        def.reject(new Error('icon ' + icon.file + ' has no name or code'));
        return false;
      }

      // 有 d 的前提下可以不写 file
      if (!options.readFiles && !icon.d) {
        def.reject(new Error('icon ' + icon.name + ' has no path data(d)'));
        return false;
      }
    });

    def.resolve(options.icons);
    return def.promise;
  } else {
    // 如果没有 icons 数据，从 src 里自动生成
    var base = options.startCodePoint;

    return Q.nfcall(fs.readdir, options.src)
      .then(function(files) {
        var svgFiles = _.filter(files, function(file) {
          return /\.svg$/i.test(file);
        });

        return _.map(svgFiles, function(file) {
          return {
            name: 'glyph-' + base,
            codepoint: base++,
            file: file
          };
        });
      });
  }
}

module.exports = builder;
